### PR TITLE
Fix config + attn_implementation in AutoModelForCausalLM.from_pretrained

### LINKED ISSUE #30298: Attention implementation cannot work together with config in AutoModel
### System Info

- `transformers` version: 4.40.0.dev0
- Platform: Linux-5.15.0-100-generic-x86_64-with-glibc2.35
- Python version: 3.11.8
- Huggingface_hub version: 0.21.4
- Safetensors version: 0.4.2
- Accelerate version: 0.28.0
- PyTorch version (GPU?): 2.2.0+cu121 (True)

### Who can help?

@younesbelkada 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

Similar to https://github.com/huggingface/transformers/issues/28038

We want to pass a model config to `from_pretrained` with an `attn_implementation` parameter. The attention type cannot be faithful to the one in the `attn_implementation`

```python
from transformers import AutoConfig, AutoModelForCausalLM
model_name = "meta-llama/Llama-2-7b-hf"
config = AutoConfig.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name, config=config, attn_implementation="eager")
print(model.config._attn_implementation)
# sdpa
```

### Expected behavior

`_attn_implementation` should be `eager`

### PR DESCRIPTION
**Pull request by hiyouga**
# What does this PR do?

Fixes #30298 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@ArthurZucker and @younesbelkada

### CODE REVIEW

**Review by fxmarty — APPROVED** [review]
This looks very reasonable, thank you for the fix. Ideally, could you add a test for this? For example, in `tests/test_modeling_utils.py`'s `TestAttentionImplementation`.

**Comment by hiyouga** [comment]
> This looks very reasonable, thank you for the fix. Ideally, could you add a test for this? For example, in `tests/test_modeling_utils.py`'s `TestAttentionImplementation`.

Ok I'll do it

**Comment by hiyouga** [comment]
@fxmarty I have done, could you review it again?

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -3145,7 +3145,7 @@ def from_pretrained(
             config = copy.deepcopy(config)
 
             kwarg_attn_imp = kwargs.pop("attn_implementation", None)
-            if kwarg_attn_imp is not None and config._attn_implementation != kwarg_attn_imp:
+            if kwarg_attn_imp is not None:
```
Comment (amyeroberts) [inline src/transformers/modeling_utils.py:3148]: Just from this diff it's not obvious at all to me why this works - when `kwarg_attn_imp` is not None. Previously: 

* Either `config._attn_implementation != kwarg_attn_imp`, in which case we set `config._attn_implementation = kwarg_attn_imp`
* Or `config._attn_implementation == kwarg_attn_imp`, in which case there's no need to set `config._attn_implementation = kwarg_attn_imp` as they're already equal. 

From the tests it works, but it's not clear why. Perhaps @fxmarty can help shed some light

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on this and fixing! 

Just a question about why this works

File: src/transformers/modeling_utils.py (hunk 1 above)
Comment (fxmarty) [reply src/transformers/modeling_utils.py:3148]: @amyeroberts a detailed explanation is here: https://github.com/huggingface/transformers/issues/30298#issuecomment-2061845587

Basically the `_attn_implementation` setter (here https://github.com/huggingface/transformers/blob/e4ea19b958c89d61e42461fac6ac8441787121f8/src/transformers/configuration_utils.py#L406-L420) is never called in this case, resulting in SDPA attention being used although it should be eager.

File: tests/test_modeling_utils.py [hunk 2]
```diff
@@ -427,6 +427,29 @@ def remove_torch_dtype(model_path):
...
+            model = AutoModelForCausalLM.from_pretrained(
+                TINY_MISTRAL, attn_implementation=requested_attn_implementation
+            )
+            self.assertEqual(model.config._attn_implementation, requested_attn_implementation)
+
+            config = AutoConfig.from_pretrained(TINY_MISTRAL)
+            model = AutoModelForCausalLM.from_pretrained(
+                TINY_MISTRAL, config=config, attn_implementation=requested_attn_implementation
+            )
+            self.assertEqual(model.config._attn_implementation, requested_attn_implementation)
```
Comment (fxmarty) [inline tests/test_modeling_utils.py:451]: Could you add a check like:
```python
for _, submodule in model.named_modules():
    if isinstance(submodule, XXXAttentionClassCorrespondingTo_attn_implementation):
    break
else:
    raise ValueError("The loaded attention class is wrong")
```

**Review by helloworld1 — APPROVED** [review]
Thanks for fixing this. I just encountered the same bug.

File: tests/test_modeling_utils.py (hunk 2 above)
Comment (hiyouga) [reply tests/test_modeling_utils.py]: Hi, I have added the check

**Review by fxmarty — APPROVED** [review]
LGTM

**Comment by fxmarty** [comment]
@hiyouga something is wrong with the CI could you try merging main?

**Review by amyeroberts — APPROVED** [review]
Thanks!

**Comment by hiyouga** [comment]
> @hiyouga something is wrong with the CI could you try merging main?

Done

**Pull request merged**